### PR TITLE
Ros2 port dev

### PR DESCRIPTION
**Pull request by mlautman**
Fixing the eval functionality and the variable length array functionality

Setting and getting parameters from a Node
rqt_service_caller_input_methods

Setting parameters explicitly and in batch
rqt_service_caller

### CODE REVIEW

**Comment by dirk-thomas** [comment]
This PR must target `crystal-devel`.

File: CMakeLists.txt [hunk 1]
```diff
@@ -1,18 +1,41 @@
...
-catkin_python_setup()
+
+find_package(ament_cmake REQUIRED)
+find_package(ament_cmake_python REQUIRED)
+find_package(rqt_py_common REQUIRED)
+find_package(rosidl_default_generators REQUIRED)
+
+set(srv_files
+  "srv/Sum.srv"
+)
```
Comment (dirk-thomas) [inline CMakeLists.txt]: This service files shouldn't be installed.

It should probably be even be removed - see other comment.

File: srv/Sum.srv [hunk 2]
```diff
@@ -0,0 +1,3 @@
+rqt_py_common/ArrayVal val
+---
+float64 total
```
Comment (dirk-thomas) [inline srv/Sum.srv]: Missing newline at the end of the file.

File: scripts/example_service [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+    rclpy.spin(node)
+
+    # Destroy the node explicitly
+    # (optional - Done automatically when node is garbage collected)
+    node.destroy_node()
+    rclpy.shutdown()
+
+
+if __name__ == '__main__':
+    main()
```
Comment (dirk-thomas) [inline scripts/example_service]: I don't think this package needs an example service script. You can easily test this with any existing demo node.

File: CMakeLists.txt [hunk 4]
```diff
@@ -1,18 +1,41 @@
...
-  DESTINATION ${CATKIN_PACKAGE_SHARE_DESTINATION}
+  DESTINATION share/${PROJECT_NAME}
 )
 
 install(DIRECTORY resource
-  DESTINATION ${CATKIN_PACKAGE_SHARE_DESTINATION}
+  DESTINATION share/${PROJECT_NAME}
+)
+
+install(PROGRAMS scripts/example_service
```
Comment (dirk-thomas) [inline CMakeLists.txt]: This shouldn't be installed.

It should probably be even be removed - see other comment.

File: src/rqt_service_caller/service_caller_widget.py [hunk 5]
```diff
@@ -196,41 +221,85 @@ def request_tree_widget_itemChanged(self, item, column):
...
-
-            self._eval_locals['i'] = counter
-            value = self._evaluate_expression(expression, slot_type)
-            if value is not None:
-                setattr(message, slot_name, value)
+        # qDebug(
+        #     'fill_message_slots('
+        #     '\n\tmessage={}, \n\ttopic_name={}, \n\texpressions={}, \n\tcounter={})'.format(
+        #         message, topic_name, expressions, counter))
+        if type(message) in (list, set) and len(message) > 0:
```
Comment (brawner) [inline src/rqt_service_caller/service_caller_widget.py]: Is there a reason to check if ```len(message) > 0```. Seems like this code path should handle the empty list/set/tuple situation

File: scripts/example_service (hunk 3 above)
Comment (mlautman) [reply scripts/example_service]: The existing demo nodes are very basic and do not test corner cases such as setting arrays, setting fields with full message definitions etc.. I'll split the example node into a separate repo

Comment (dirk-thomas) [reply scripts/example_service]: Why can't you use one of the existing demo nodes but need your own custom one?

File: CMakeLists.txt [hunk 6]
```diff
@@ -4,17 +4,6 @@ project(rqt_service_caller)
 find_package(ament_cmake REQUIRED)
 find_package(ament_cmake_python REQUIRED)
 find_package(rqt_py_common REQUIRED)
```
Comment (dirk-thomas) [inline CMakeLists.txt]: This package is not a build dependency and therefore not available at build time. See #8.

File: CMakeLists.txt [hunk 7]
```diff
@@ -4,17 +4,6 @@ project(rqt_service_caller)
...
-find_package(rosidl_default_generators REQUIRED)
-
-set(srv_files
-  "srv/Sum.srv"
```
Comment (dirk-thomas) [inline CMakeLists.txt]: The `.srv` file wasn't deleted: see #8.
